ts/private/ts_proto_library.bzl

```diff
@@ -0,0 +1,116 @@
+"Private implementation details for ts_proto_library"
+
+load("@aspect_rules_js//js:providers.bzl", "js_info")
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load("@rules_proto//proto:defs.bzl", "ProtoInfo")
+load(":ts_lib.bzl", ts_lib = "lib")
+
+def _short_path(f):
+    return f.short_path
+
+# buildifier: disable=function-docstring-header
+def _protoc_action(ctx, inputs, descriptors, outputs, options = {
+    "keep_empty_files": True,
+    "target": "js+dts",
+}):
+    """Create an action like
+    bazel-out/k8-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/protoc $@' '' \
+      '--plugin=protoc-gen-es=bazel-out/k8-opt-exec-2B5CBBC6/bin/plugin/bufbuild/protoc-gen-es.sh' \
+      '--es_opt=keep_empty_files=true' '--es_opt=target=ts' \
+      '--es_out=bazel-out/k8-fastbuild/bin' \
+      '--descriptor_set_in=bazel-out/k8-fastbuild/bin/external/com_google_protobuf/timestamp_proto-descriptor-set.proto.bin:bazel-out/k8-fastbuild/bin/example/thing/thing_proto-descriptor-set.proto.bin:bazel-out/k8-fastbuild/bin/example/place/place_proto-descriptor-set.proto.bin:bazel-out/k8-fastbuild/bin/example/person/person_proto-descriptor-set.proto.bin' \
+      example/person/person.proto
+    """
+    args = ctx.actions.args()
+    args.add_joined(["--plugin", "protoc-gen-es", ctx.executable.protoc_gen_es.path], join_with = "=")
+    for (key, value) in options.items():
+        args.add_joined(["--es_opt", key, value], join_with = "=")
+    args.add_joined(["--es_out", ctx.bin_dir.path], join_with = "=")
+
+    if ctx.attr.has_services:
+        args.add_joined(["--plugin", "protoc-gen-connect-es", ctx.executable.protoc_gen_connect_es.path], join_with = "=")
+        for (key, value) in options.items():
+            args.add_joined(["--connect-es_opt", key, value], join_with = "=")
+        args.add_joined(["--connect-es_out", ctx.bin_dir.path], join_with = "=")
+
+    args.add("--descriptor_set_in")
+    args.add_joined(descriptors, join_with = ":")
+
+    args.add_all(inputs)
+    ctx.actions.run(
+        executable = ctx.executable.protoc,
+        progress_message = "Generating .js/.d.ts from %{label}",
+        outputs = outputs,
+        inputs = depset(inputs, transitive = [descriptors]),
+        arguments = [args],
+        tools = [ctx.executable.protoc_gen_es] + (
+            [ctx.executable.protoc_gen_connect_es] if ctx.attr.has_services else []
+        ),
+        env = {"BAZEL_BINDIR": ctx.bin_dir.path},
+    )
+
+def _declare_outs(ctx, proto_srcs, ext):
```

Comment:
use `proto_common.declare_generated_files` instead. https://github.com/bazelbuild/bazel/blob/9573fcf095279e76ef06fcfff90a1188c944f381/src/main/starlark/builtins_bzl/common/proto/proto_common.bzl#L351. which can be imported from rules_proto defs https://github.com/bazelbuild/rules_proto/blob/493169c1199dc21b9da860f7040a4502aa174676/proto/defs.bzl#L63C1-L63C1